swap/src/asb/event_loop.rs

```diff
@@ -319,13 +320,40 @@ where
         min_buy: bitcoin::Amount,
         max_buy: bitcoin::Amount,
     ) -> Result<BidQuote> {
+        let monero_decimal = self.monero_wallet.get_balance().await?.as_monero_decimal();
+
         let rate = self
             .latest_rate
             .latest_rate()
-            .context("Failed to get latest rate")?;
+            .context("Failed to get latest rate")?
+            .ask()
+            .context("Failed to compute asking price")?;
+
+        let bitcoin_decimal: Decimal = Decimal::from(rate.as_sat()) / Decimal::from(100_000_000);
+
+        let max_bitcoin_for_monero = match (monero_decimal * bitcoin_decimal).to_f64() {
+            Some(val) => bitcoin::Amount::from_btc(val)?,
+            None => {
+                tracing::error!("There was a problem calculating the max amount for Bitcoin sell");
+                tracing::warn!("Using Max purchase from config: {}", max_buy);
+                max_buy
+            }
+        };
+
+        if max_buy > max_bitcoin_for_monero {
+            tracing::warn!(
+                    "Your Max bitcoin to sell config: {} is bigger than your current balance of Monero available to sell {}",
+                    max_buy, max_bitcoin_for_monero
+                );
```

Comment:
Can you please extract this calculation into a dedicated function or even better, move it _onto_ the actual types itself?

There is a lot of low-level logic happening here and we don't have any tests for it.

Reply:
If you look at what the function did before, then you can see that it only gathered some data and combined it into a `BidQuote` but the actual calculation of what an _ask_ price is, does not happen in this function.

It would like to it stay that way. In other words: The responsibility of `make_quote` should be to construct a `BidQuote` by gathering data from several dependencies (like wallets, rates etc). The actual calculation of the price is better extracted into a pure function that does the computation based on the inputs. That function we can then test for all its edge-cases.